File: CodingKatas/old/katas/dateidoubletten/impl_java/DublettenPruefung.java
package katas.dateidoubletten.impl_java;

import katas.dateidoubletten.IDublette;
import katas.dateidoubletten.IDublettenpruefung;
import katas.dateidoubletten.Vergleichsmodi;

import java.io.IOException;
import java.math.BigInteger;
import java.nio.file.FileSystem;
import java.nio.file.Files;
import java.nio.file.Path;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Stream;

public class DublettenPruefung implements IDublettenpruefung {
    private final FileSystem fileSystem;
    public DublettenPruefung(FileSystem fileSystem) {
        this.fileSystem = fileSystem;
    }
    @Override
    public List<IDublette> sammleKandidaten(String pfad) {
        return sammleKandidaten(pfad, Vergleichsmodi.GROESSE_UND_NAME); // default
    }

    @Override
    public List<IDublette> sammleKandidaten(String pfad, Vergleichsmodi modus) {
        return sammleKandidaten(fileSystem.getPath(pfad), modus);
    }

    private List<IDublette> sammleKandidaten(Path pPath, Vergleichsmodi modus) {
        DublettenSammler dublettenSammler = new DublettenSammler();
        try (Stream<Path> walk = Files.walk(pPath)) {
            walk.filter(Files::isRegularFile).forEach( filePath ->
                    dublettenSammler.addPathWithHash(filePath, getHash(filePath, modus)));
        } catch (IOException ex) {
            throw new RuntimeException(ex);
        }
        return dublettenSammler.getDubletten();
    }

    private String getHash(Path path, Vergleichsmodi modus) {
        try {
            return switch (modus) {
                case GROESSE_UND_NAME -> "hash_" + path.getFileName() + "_" + Files.size(path);
                case GROESSE ->          "hash__" + Files.size(path);
                default -> "";
            };
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public List<IDublette> pruefeKandidaten(List<IDublette> kandidaten) {
        List<IDublette> dubletten = new ArrayList<>();
        kandidaten.forEach(kandidat -> {
            List<String> dateipfade = kandidat.dateipfade();
            dubletten.addAll(pruefeKandidaten(dateipfade, new DublettenSammler()));
        });
        return dubletten;
    }

    private List<IDublette> pruefeKandidaten(List<String> dateipfade, DublettenSammler dublettenSammler) {
        dateipfade.forEach(dateipfad -> {
            Path path = fileSystem.getPath(dateipfad);
            dublettenSammler.addPathWithHash(path, getMD5Hash(path));
        });
        return dublettenSammler.getDubletten();
    }

    private String getMD5Hash(Path filePath) {
        try {
            byte[] data = Files.readAllBytes(filePath);
            byte[] hash = MessageDigest.getInstance("MD5").digest(data);
            return new BigInteger(1, hash).toString(16);
        } catch (IOException | NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
    }
}

File: CodingKatas/old/katas/dateidoubletten/impl_kotlin/Dublettenpruefung.kt
package katas.dateidoubletten.impl_kotlin

import katas.dateidoubletten.IDublette
import katas.dateidoubletten.IDublettenpruefung
import katas.dateidoubletten.Vergleichsmodi
import java.io.File
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import kotlin.io.path.Path
import kotlin.io.path.name
import kotlin.io.path.pathString

class Dublettenpruefung : IDublettenpruefung {


    override fun sammleKandidaten(pfad: String): List<IDublette> {
        return sammleKandidaten(pfad, Vergleichsmodi.GROESSE_UND_NAME)
    }

    override fun sammleKandidaten(pfad: String, modus: Vergleichsmodi): List<IDublette> {
        return sammleKandidaten(Path(pfad), modus, HashMap<String, MutableList<Path>>())
    }

    private fun sammleKandidaten(pPath: Path, modus: Vergleichsmodi, map: java.util.HashMap<String, MutableList<Path>>): List<IDublette> {
        return try {
            File(pPath.name).walkTopDown().forEach {
                println(it)
                if (it.isFile) {
                    addPathWithHashToMap(map, it.toPath(), getHash(it.toPath(), modus))
                }
                if (it.isDirectory && !it.path.equals(pPath.pathString)) {
                    sammleKandidaten(it.toPath(), modus, map) // recursion
                }
            }
            return getDublettenFromMap(map)
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
    }

    private fun getDublettenFromMap(map: java.util.HashMap<String, MutableList<Path>>): List<IDublette> {
        /*
        return map.values.stream()
            .filter { list -> list.size > 1 }
            .map {Dublette(it)}
            .toList()

         */
        return emptyList()
    }

    private fun addPathWithHashToMap(map: java.util.HashMap<String, MutableList<Path>>, path: Path, hash: String) {
        val hashFiles = map.computeIfAbsent(hash) { l: String -> mutableListOf() }
        hashFiles.add(path)
    }

    private fun getHash(path: Path, modus: Vergleichsmodi): String {
        return try {
            when (modus) {
                Vergleichsmodi.GROESSE_UND_NAME -> {
                    "hash_" + path.fileName + "_" + Files.size(path)
                }
                Vergleichsmodi.GROESSE -> {
                    "hash__" + Files.size(path)
                }
            }
        } catch (e: IOException) {
            throw java.lang.RuntimeException(e)
        }
    }

    override fun pruefeKandidaten(kandidaten: List<IDublette>?): List<IDublette> {
        TODO("Not yet implemented")
    }
}

fun main() {
    val dp = Dublettenpruefung()
    val sammleKandidaten = dp.sammleKandidaten("C:\\workspace\\content-service\\src")
    println(sammleKandidaten)
}
